src/commonMain/kotlin/fr/acinq/lightning/crypto/KeyManager.kt

```diff
@@ -118,28 +122,30 @@ interface KeyManager {
         val refundDelay: Int = DefaultSwapInParams.RefundDelay
     ) {
         private val userExtendedPrivateKey: DeterministicWallet.ExtendedPrivateKey = DeterministicWallet.derivePrivateKey(master, swapInUserKeyPath(chain))
+        private val userRefundExtendedPrivateKey: DeterministicWallet.ExtendedPrivateKey = DeterministicWallet.derivePrivateKey(master, swapInUserRefundKeyPath(chain))
+
         val userPrivateKey: PrivateKey = userExtendedPrivateKey.privateKey
         val userPublicKey: PublicKey = userPrivateKey.publicKey()
 
+        val userRefundPrivateKey: PrivateKey = DeterministicWallet.derivePrivateKey(userRefundExtendedPrivateKey, 0).privateKey
+        val userRefundPublicKey: PublicKey = userRefundPrivateKey.publicKey()
+
         private val localServerExtendedPrivateKey: DeterministicWallet.ExtendedPrivateKey = DeterministicWallet.derivePrivateKey(master, swapInLocalServerKeyPath(chain))
         fun localServerPrivateKey(remoteNodeId: PublicKey): PrivateKey = DeterministicWallet.derivePrivateKey(localServerExtendedPrivateKey, perUserPath(remoteNodeId)).privateKey
 
-        val redeemScript: List<ScriptElt> = Scripts.swapIn2of2(userPublicKey, remoteServerPublicKey, refundDelay)
-        val pubkeyScript: List<ScriptElt> = Script.pay2wsh(redeemScript)
-        val address: String = Bitcoin.addressFromPublicKeyScript(chain.chainHash, pubkeyScript).result!!
+        val swapInProtocol = SwapInProtocol(userPublicKey, remoteServerPublicKey, userRefundPublicKey, refundDelay)
```

Comment:
`SwapInOnChainKeys` has a `chain` attribute, but when we do `keyManager.swapInProtocol` we get an object that is chain-agnostic, and need to provide the chain to `swapInProtocol.address()` and `swapInProtocol.descriptor()`. Why not add `chain` to `SwapInProtocol`?

Reply:
I've reworked the generation of descriptors and our recovery procedure (which had problems, see the commit message). There are now 2 different methods, to generate either a private descriptor or a public descriptor. Both are valid, compatible with refund key rotation, and can be imported directly into bitcoin core.

For the recovery process, the user's seed is not enough (it's true for the current swaps too): users also need their swap-in public key, the server's swap-in public key, and the refund delay.

Reply:
see https://github.com/ACINQ/lightning-kmp/pull/563/commits/689949a9a2c6bf9e86de9d2db34f1c686a85e66a for more info

Reply:
An example of that is that the `compute descriptors to recover swap-in funds` test isn't a very good test: it uses custom code in the test itself instead of testing the actual code paths used by a wallet user. It would be better to use the seeds for the standard test users (`TestConstants.Alice` and `TestConstants.Bob`) and the corresponding `keyManager` instead to derive the descriptors?

Reply:
> I don't understand why you moved the `descriptor` functions to the `companion object`: you need to provide the same parameters as the `(Legacy)SwapInProtocol` instance, so they should just be methods? Otherwise you're risking that we mess up the parameters and don't provide the right ones (or at least it's just annoying because we have to provide a list of parameters we already provided when creating the `(Legacy)SwapInProtocol` instances)...

in the legacy case `SwapInProtocolLegacy` has all that is needed to generate a descriptor. 
But for the new swap-ins, you need to provide the refund master xpriv, and the descriptor will be valid only if refund keys passed to create instances of `SwapInProtocol` are derived directly from this xpriv, so I think that having a static method that takes all required parameters is a bit cleaner.

Reply:
I'm not really convinced to be honest, making it a method lets us provide less arguments and makes it easy to check correctness of the arguments provided inside the function with a `require`, but that's not very important, you can ignore that.